### PR TITLE
[runtime env] Parse local pip/conda requirements files locally upon task/actor definition

### LINKED ISSUE #18976: [runtime env] [Bug] Specifying pip "requirements.txt" per-task doesn't work on remote cluster
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Core

### What happened + What you expected to happen

Specifying pip requirements per-task in a local `requirements.txt` file doesn't work when using a remote cluster. It tries to parse the `requirements.txt` file on the cluster side, but the file doesn't exist there.  This bug is present in 1.7.0rc0 and 1.6.0, and possibly earlier versions.

### Reproduction script

 Create a folder with a `requirements.txt` file containing the single line `pip-install-test` and navigate to that folder.

Start a remote cluster with `ray up -y config.yaml`.  I used the Cluster launcher with the following config.yaml:
```yaml
# An unique identifier for the head node and workers of this cluster.
cluster_name: minimal-170rc0-archit

# Cloud-provider specific configuration.
provider:
    type: aws
    region: us-west-2

setup_commands:
    - pip uninstall -y ray && pip install -U "ray[default]==1.7.0rc0"
```

Then run the following repro script:
```python
import ray
ray.init("ray://54.70.131.201:10001")

@ray.remote 
def test_import(): 
    import pip_install_test
    return True            

assert ray.get( 
    test_import.options(runtime_env={"pip": "requirements.txt"}).remote()
) 
```

We get the following error:
```python-traceback
~/anaconda3/envs/17rc0-2/lib/python3.7/site-packages/ray/util/client/worker.py in _call_schedule_for_task(self, task)
    468         if not ticket.valid:
    469             try:
--> 470                 raise cloudpickle.loads(ticket.error)
    471             except (pickle.UnpicklingError, TypeError):
    472                 logger.exception("Failed to deserialize {}".format(

ValueError: requirements.txt is not a valid file
```
### Anything else

This occurs whether you use `requirements.txt`, `./requirements.txt`, or the full absolute path.

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by architkulkarni**
## Why are these changes needed?
When using Ray Client, the `.remote()` function for actors and tasks are run on the local driver and run a second time on the remote Client server.

Previously, runtime env dict validation took place in `.remote()`.  This caused a bug where a local file path was specified inthe `"pip"` or `"conda"` field, because when the validation was run in the Client server, the local file wouldn't be present.

This PR fixes this by separating out the `pip` and `conda` validation and moving it to the `__init__` and `.options()` methods for actors and tasks.  Here when a filename string is given, it is parsed into the internal representation `List[str]` (for `pip`) or `Dict` (for `conda`).  Even though `.options()` is called a second time on the client server, the validation is a no-op there because the `conda` and `pip` fields are already in the internal representation at that point.

## Related issue number

Closes #18976 
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by architkulkarni** [comment]
@edoakes A couple issues here.

1. It looks like the `client_convert` function reruns the entire `_remote()` function, including the runtime env rewriting part that we just added above the `client_convert` call.  In particular, that rewriting part gets called twice, once locally and once in the client server.  Hence we see the following error:
```
ValueError: pip-install-test
requests==2.24.0
 is not a valid file
```
because it already got rewritten from filename -> file contents once, and during the second call it tries to interpret the file contents as a filename.

2. Since our changes are still happening in `_remote()`, it's not clear how to locally test it, since the validation hasn't been moved to the `init()` or `.options()` part.

File: python/ray/actor.py [hunk 1]
```diff
@@ -388,11 +389,13 @@ class DerivedActorClass(cls, modified_class):
             PythonFunctionDescriptor.from_class(
                 modified_class.__ray_actor_class__)
 
+        new_runtime_env = parse_pip_and_conda(runtime_env)
```
Comment (edoakes) [inline python/ray/actor.py:392]: comment about why we do it here

Comment (edoakes) [reply python/ray/actor.py:392]: (and other places)

**Review by edoakes — APPROVED** [review]
Looks good to me, hopefully this works :)

**Comment by architkulkarni** [comment]
This PR unfortunately does not address https://github.com/ray-project/ray/issues/19002, which I think probably has an unrelated root cause.

**Pull request merged**